src/datatypes.jl

```diff
@@ -308,6 +308,7 @@ function Datatype(::Type{T}; commit=true) where T
     if commit
         commit!(dt)
     end
+    @eval Datatype(::Type{$T}) = $dt
```

Comment:
While technically legal it is not a great idea, you can run into `worldage` issues with this.

if you need caching, I would go through a global `IDDict{Type, Dataype}`.

Reply:
Maybe we should just make `Datatype(T)` a generated function?

Reply:
Thanks for the comments! I'm still learning about world age, but now I see how this can be a problem. So one would have to instruct users to once call `MPI.Datatype(T)` at top level, in order to be able to take advantage of the new method later. That would not be elegant.

It seems to me that performance should be much better than with dictionary lookup, though, which presumably would have to be done at runtime on every `Put`.

Making `Datatype()` a generated function sounds like a good solution, as I understand that generated functions can cache the generated (and compiled) code.